include/RAJA/pattern/forall.hpp

```diff
@@ -379,13 +458,28 @@ forall_Icount(ExecutionPolicy&& p,
  *
  ******************************************************************************
  */
-template <typename ExecutionPolicy, typename Container, typename LoopBody>
-RAJA_INLINE concepts::enable_if<
+template <typename ExecutionPolicy, typename Container, typename LoopBody, typename Res>
+RAJA_INLINE concepts::enable_if_t<
+    resources::EventProxy<Res>,
     concepts::negate<type_traits::is_indexset_policy<ExecutionPolicy>>,
     concepts::negate<type_traits::is_multi_policy<ExecutionPolicy>>,
     type_traits::is_range<Container>>
 forall(ExecutionPolicy&& p, Container&& c, LoopBody&& loop_body)
 {
+  std::cout<< "Check10\n";
+  auto r = resources::get_default_resource(p);
+  return forall(r, p, c, loop_body);
+}
+
+template <typename Res, typename ExecutionPolicy, typename Container, typename LoopBody>
+RAJA_INLINE concepts::enable_if_t<
+    resources::EventProxy<Res>,
+    concepts::negate<type_traits::is_indexset_policy<ExecutionPolicy>>,
+    concepts::negate<type_traits::is_multi_policy<ExecutionPolicy>>,
+    type_traits::is_range<Container>>
+forall(Res &r, ExecutionPolicy&& p, Container&& c, LoopBody&& loop_body)
```

Comment:
This is the line I should have picked.

Reply:
I see your point. Something I'm also currently thinking about is how to pass resources for indexset policies and multipolicies, since they could use multiple resources i.e. host and cuda for an indexset or any number of resources for a given multipolicy, how exactly should that interface look to the user, and more importantly how to implement it.

Reply:
@mdavis36 for IndexSets, I think we want to keep things simple. Here are the combinations I think we need to support.

1. Host only execution: Outer iteration over segments, either sequential or threaded (e.g., OpenMP),  and individual segment execution either sequential or threaded. But, don't thread both. We don't support this (yet) and it's not clear how well OpenMP teams would work with this.
2. For GPU execution, only the segments will execute in GPU kernels and the outer iteration over segments will always happen on the host. The host segment iteration could use sequential or threaded execution.

If you think only about these possibilities, does it simplify things?

Reply:
It does thank you I didn't know we only support IndexSets in those configurations, but it makes sense. I think something else I'm interested in for this PR is how we want the user to pass the arguments for multiple resources. Do we want to pass them directly and use some argument unpacking magic to handle them like this:
- IndexSet Policy - `forall(IndexSetExecPol, ResourceIter, ResourceExec, Container ... )`
- MultiPolicy - `forall(MultiPolExecPol, Res1, Res2, Res3, Res4, Container .... )`

Or would it be better to have the user pass them in like a tuple and pull the resources out internally so it would look like this:
- IndexSet Policy - `forall(IndexSetExecPol, MultiResTuple<ResourceIter, ResourceExec>, Container ... )`
- MultiPolicy - `forall(MultiPolExecPol, MultiResTuple<Res1, Res2, Res3, Res4>, Container .... )`

Reply:
Not sure about the multi-policy stuff since I don't really know who's using it and how. @ajkunen may have some ideas about that.

For index set stuff, do we need a resource for the outer segment iteration part since it will always be on the host? Are you thinking that one may want to stage multiple kernels that use index sets and execute them asynchronously?  

For the segment execution part, we probably should focus on the case where there are multiple segments and we want to lunch them in separate GPU streams to overlap execution as much as possible. So, we would need a resource for each segment, which could get clumsy for users. What if we just define a resource under the covers based on the execution policy?

Reply:
> For index set stuff, do we need a resource for the outer segment iteration part since it will always be on the host? Are you thinking that one may want to stage multiple kernels that use index sets and execute them asynchronously?
>

I guess we wouldn't really _need_ to define a resource for segment iteration right now, currently even if a user were to execute with an omp policy as the iterator it would still utilize the default Host resource. I was thinking this would be more of a syntactic formality for the time until we add some level of support for host side threaded/async resource execution, which would require a user to pass a specific resource object.
 
> For the segment execution part, we probably should focus on the case where there are multiple segments and we want to lunch them in separate GPU streams to overlap execution as much as possible. So, we would need a resource for each segment, which could get clumsy for users. What if we just define a resource under the covers based on the execution policy?

That could be another way to approach this, I had considered the idea of allowing a user to execute each segment on a different stream by generating our own GPU resources internally. I assumed the default behaviour would be to execute every segment on a single stream. I imagine users would define which behaviour to invoke with something in the Execution Policy. I can definitely see this being a good solution for how to handle indexset forall calls when no Resource is passed.

Reply:
We might, uses for having the outer loop be on the GPU have come up a few times, but I'm really not sure how best to handle that yet.